research/train.py

```diff
@@ -37,20 +44,59 @@ def _update(engine, batch):
         loss = loss_fn(y_pred, y)
         loss.backward()
         optimizer.step()
-        return loss.item(), y_pred, y
 
-    def _metrics_transform(output):
-        return output[1], output[2]
+        # Anything you want to log must be returned in this dictionary
+        update_dict = {"nll": loss.item(), "y_pred": y_pred, "y": y}
+        return update_dict
 
     engine = Engine(_update)
 
-    for name, metric in metrics.items():
-        metric._output_transform = _metrics_transform
-        metric.attach(engine, name)
+    # Common metrics require base "y_pred" and "y". Lambda to reduce verbosity in metrics
+    _ypred_y = lambda _: (_["y_pred"], _["y"])
+
+    # (Optional) Specify training metrics. "output_transform" used to select items from "update_dict" needed by metrics
+    # Collecting metrics over training set is not recommended
+    # https://pytorch.org/ignite/metrics.html#ignite.metrics.Loss
 
     return engine
 
 
+def create_supervised_evaluator(
+    model: nn.Module, cfg: DictConfig, device="cpu"
+) -> Engine:
+
+    # Loss
+    loss_fn = torch.nn.NLLLoss()
+
+    def _inference(engine, batch):
+        model.eval()
+        with torch.no_grad():
+            x, y = batch
+            x, y = x.to(device), y.to(device)
+            y_pred = model(x)
+
+        # Anything you want to log must be returned in this dictionary
+        infer_dict = {"y_pred": y_pred, "y": y}
```

Comment:
From what I can tell this means that all outputs will be held in the GPU until the entire evaluation is finished since there will be an active reference to this data. The garbage collector can't collect it until after the entire evaluation finishes. This is potentially problematic since you will have growing memory usage. There can be large evaluation datasets. What is done sometimes instead is that you record the metrics per batch and at the end of the batch you combine all the metrics into the final metric (not necessarily an average since probabilties can't be averaged to get the correct result).

This allows you to deference and free memory usage for the next batch while still having computed a metric

Reply:
Couldnt you just do y_pred.detach().cpu() or smth like this

Reply:
Especially cos sometimes you actually  want to save  the predictions, not just the metrics, to look through them for understanding

Reply:
@ChickenTarm Looking at the documentation, I don't think that's what happens. Outputs from `_inference` are passed into whatever metrics you create (e.g. `Accuracy`) or we log them. I.e. metrics are computed **online**, we aren't accumulating all the `y` and `y_pred`, then computing metrics.

Reply:
Taken from the footnote, we are accumulating the metrics online in the evaluation set:

https://pytorch.org/ignite/quickstart.html#code

`Another more common pattern is to accumulate measures online over an epoch in the training loop. In this case metrics are aggregated on a moving model, and thus, we do not want to encourage this pattern.`

Reply:
@mchiquier you can detach and send to cpu but that would slow down inference time. since transferring data is some of the most expensive operations. But that also assumes that the machine has enough RAM to store the entire y_pred which might not be the case.

@goodmattg Okay then that checks out looks good